### PR TITLE
Do not save methods as properties

### PR DESCRIPTION
**Pull request by samo98**
* In type collection save methods only as methods

### CODE REVIEW

File: Sources/Fuzzilli/FuzzIL/TypeSystem.swift [hunk 1]
```diff
@@ -590,20 +590,23 @@ public struct Type: Hashable {
         guard Is(.object()) else {
             return self
         }
-        var newProperties = properties
+        var newProperties = properties, newMethods = methods
         newProperties.insert(property)
-        let newExt = TypeExtension(group: group, properties: newProperties, methods: methods, signature: signature)
+        // We are adding proprety which could have been method in the past
```
Comment (saelo) [inline Sources/Fuzzilli/FuzzIL/TypeSystem.swift:595]: "proprety" -> "a property"

File: Sources/Fuzzilli/FuzzIL/TypeSystem.swift [hunk 2]
```diff
@@ -612,20 +615,11 @@ public struct Type: Hashable {
...
-    
-    /// Returns a new ObjectType that represents this type without the removed property.
-    public func removing(method: String) -> Type {
-        guard Is(.object()) else {
-            return self
-        }
-        var newMethods = methods
-        newMethods.remove(method)
-        let newExt = TypeExtension(group: group, properties: properties, methods: newMethods, signature: signature)
+        // We are adding method which could have been property in the past
```
Comment (saelo) [inline Sources/Fuzzilli/FuzzIL/TypeSystem.swift:620]: Nit: "a method"

**Review by saelo — COMMENTED** [review]
Yep, nice!
So I'm now generally wondering how we should treat properties and methods in Fuzzilli. Your suggestion here definitely makes sense, i.e. properly separating the two in the type system and type collection. I'm thinking maybe we should to that in other parts of Fuzzilli as well:
- Introduce two new operations: StoreMethod and LoadMethod which are basically like {Store/Load}Property but access the type.methods set and require a .function() as 2nd input. 
    - This will allow easy tuning of their frequency in the generated code and will make reasoning about the code more or less trivial.
    - Also, this allows modelling "function properties", which are really just functions (not methods) but stored as a property on an object (they would then most of the time be loaded into a local variable and then be called without a reference object)
    - It would allow for a "writeMethodNames" set in the environment, which would contain e.g. `valueOf` (currently a propertyName: https://github.com/googleprojectzero/fuzzilli/blob/master/Sources/Fuzzilli/Core/JavaScriptEnvironment.swift#L166) and thus ensure that valueOf would be a valid method (otherwise it's invalid)
    - Finally, it would result in most property loads actually loading a property, and not a method (as they probably do now since there are just way more methods than properties)
- Separate the propertyNames and methodNames sets in the environment as well (I think this will not require extra work)
- In the future, this will allow us to have more powerful ObjectLiterals, with things like methods, getters, and setters defined inline. This could be quite interesting to have
- We can probably still leave the "optimization" in the interpreter that a property value of type .function is also a method. Alternatively, the interpreter can just ignore "invalid" StoreProperty and StoreMethod occurrences (e.g. StoreMethod where the value is not a function). Probably this doesn't really matter much
WDYT?

File: Tests/FuzzilliTests/TypeSystemTest.swift [hunk 3]
```diff
@@ -428,9 +428,9 @@ class TypeSystemTests: XCTestCase {
         XCTAssertEqual(fooObj.adding(method: "baz"), fooBazObj)
         XCTAssertEqual(bazObj.adding(method: "foo"), fooBazObj)
         
-        XCTAssertEqual(fooBarObj.removing(method: "baz"), fooBarObj)
-        XCTAssertEqual(fooBarObj.removing(method: "foo"), barObj)
-        XCTAssertEqual(barObj.removing(method: "bar"), object)
+        XCTAssertEqual(fooBarObj.removing(property: "baz"), fooBarObj)
```
[the remark this answers is not in the document]
Comment (samo98) [reply Tests/FuzzilliTests/TypeSystemTest.swift:431]: Until now methods were saved also under properties so I wanted to have single function to remove both cases and did not find better naming (actually method is also property in this context)

But as we plan totally separating them (introducing new operation and so on), then yeah I will keep both types of functions (property removing and method removing)

[the remark this answers is not in the document]
Comment (saelo) [reply Tests/FuzzilliTests/TypeSystemTest.swift:431]: Ok, sounds good!

**Pull request merged**